### PR TITLE
Dynamic Batch Sizes

### PR DESCRIPTION
**Pull request by jmhessel**
### Summary

Attempt to support dynamic batch sizes by giving API access to the `check_array_lengths` parameter. See https://github.com/keras-team/keras/issues/11993.

`model.predict` already works with dynamic batch sizes, but this PR makes training work too via:

`model.compile( ... , check_array_lengths=False)`

### Related Issues

https://github.com/keras-team/keras/issues/11993

### PR Overview

- [y] This PR requires new unit tests [y/n] (make sure tests are included)
- [y] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [n] This PR is backwards compatible [y/n]
- [y] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

PR is not backwards compatible because it doesn't work on the CNTK backend. I don't think CNTK supports multiple dynamic axes, which arise frequently in code enabled by this PR.

### CODE REVIEW

**Comment by jmhessel** [comment]
This feature can't be supported by CNTK I think right now. I am not a CNTK expert, but according to:

https://cntk.ai/pythondocs/sequence.html

> At the time of this writing CNTK supports a single additional dynamic axis.

This PR requires support for multiple dynamic axes, i.e., in addition to the batch size axis, other dimensions may be defined at execution time, too.

File: keras/engine/training.py [hunk 1]
```diff
@@ -84,21 +85,36 @@ def compile(self, optimizer,
                 can specify them via the `target_tensors` argument. It can be
                 a single tensor (for a single-output model), a list of tensors,
                 or a dict mapping output names to target tensors.
+            check_array_lengths: By default, Keras will ensure that all input
+                and output tensors have a matching batch size for each batch.
+                If you would like to skip this check and allow variable batch
+                sizes among input/output tensors, this parameter can be set
+                to True.
```
Comment (pstjohn) [inline keras/engine/training.py]: I would think `check_array_lengths` should be `False` to skip the check

Comment (jmhessel) [reply keras/engine/training.py]: Yep --- you're right. I will update the doc

Comment (jmhessel) [reply keras/engine/training.py]: Thanks for taking a look at this, BTW :)

**Comment by jmhessel** [comment]
All tests are passing now --- I'm happy to change things around if need be!

**Comment by gabrieldemarmiesse** [comment]
Since this is an API change, I'll tag @fchollet

**Comment by fchollet** [comment]
Thanks for the PR. However, I don't think we should do something like this.

- The 1:1 mapping between inputs and targets is an important assumption and we should not break it
- Implementation details should not be part of the public API

It isn't clear at all what it means to train a model when you don't have a 1:1 mapping between inputs and targets/outputs, in the general case. If a user wants to structure training in that way, that is on them, and they should write their own training loop.

**Comment by jmhessel** [comment]
Okay -- that makes sense!! I'm glad there is a solution for those of us with models that have variable input output sizes, but I guess it's too niche to justify an API change

**Comment by jmhessel** [comment]
Though maybe this should be documented somewhere? Because currently this parameter is inaccessible without overriding Model entirely which doesn't seem great, especially because everything works perfectly for my use case after knowing about that parameter. I'll think about where it could be documented.

**Comment by pstjohn** [comment]
Here's one example of a situation where you'd need this kind of functionality,
https://arxiv.org/abs/1806.01261
https://github.com/deepmind/graph_nets

The deepmind team used sonnet to implement their networks, but after my PR (#11548) you can do the same thing in keras. (i.e. https://github.com/NREL/nfp)

But you could imagine a similar situation in any sort of recurrent neural network where inputs had a variable length. The most common option is to zero-pad input matrices (here the batch size is 4)
```
input = 
[a b 0 0 0,
 c d e 0 0,
 f g 0 0 0,
 h i j k l]

output = [1.1, 1.2, 1.3, 1.4]
```
But another is to unstack the rows and pass a separate tensor that tells the network how to group them
```
input1 = [a b c d e f g h i j k l]
input2 = [1 1 2 2 2 3 3 4 4 4 4 4]

output = [1.1, 1.2, 1.3, 1.4]
```
So you can use things like `tf.segment_sum` etc to reduce back to the original batch axis.

So it's still a 1:1 mapping, just a different input scheme. Ultimately I'm assuming anyone trying to do this is going to have to write their own Keras layers to handle those odd inputs, so there's certainly merit to the argument that these checks shouldn't be *too* easy to disable

**Comment by jmhessel** [comment]
Adding to @pstjohn --- there are quite a few use-cases for dynamic input-output sizes. My use-case is: I am passing N images and M sentences, and the output is a single loss for each image and each sentence. The way I am handling this currently is to have 2 inputs of size `(None, image_feature_length)` and `(None, text_feature_length)`, except the "None's" are different lengths at run time. This couldn't be solved by writing my own training loop, because even `train_on_batch` forces the `check_array_lengths` check. This is quite similar to [this stackoverflow issue](https://stackoverflow.com/questions/44101692/implementing-a-multi-input-model-in-keras-each-with-a-different-sample-sizes-ea)):

Multi Input

It's not paramount that this be accessible through the API, but I did burn a few hours finding @pstjohn 's solution in https://github.com/keras-team/keras/pull/11548. Furthermore, you could probably solve these issues by adding additional tensor axes and padding (originally --- I had quite complicated code that masked inputs, outputs, and losses) but it's less clean.
